src/deploy/missions/base-deploy-mission.ts

```diff
@@ -0,0 +1,113 @@
+import { Contract } from "ethers";
+import { TDeployArgs, IContractDbObject, IProxyData, IDeployMissionArgs } from "./types";
+import { DeployCampaign } from "../campaign/deploy-campaign";
+import { IDeployCampaignConfig } from "../campaign/types";
+
+
+// TODO dep:
+//    1. add better logging for each step
+//    2. add proper error handling
+export class BaseDeployMission {
+  contractName! : string;
+  instanceName! : string;
+  proxyData! : IProxyData;
+  campaign : DeployCampaign;
+  logger : Console;
+  config : IDeployCampaignConfig;
+
+  constructor ({
+    campaign,
+    logger,
+    config,
+  } : IDeployMissionArgs) {
+    this.campaign = campaign;
+    this.logger = logger;
+    this.config = config;
+  }
+
+  async getFromDB () {
+    return this.campaign.dbAdapter.getContract(this.contractName);
+  }
+
+  async pushToDB (contract : Contract) {
+    const implAddress = this.proxyData.isProxy
+      ? await this.campaign.deployer.getProxyImplAddress(contract.address)
+      : null;
+
+    const contractDbDoc = this.buildDbObject(contract, implAddress);
+
+    return this.campaign.dbAdapter.writeContract(this.contractName, contractDbDoc);
+  }
+
+  async preDeploy () {
+    return Promise.resolve();
+  }
+
+  async needsDeploy () {
+    const dbContract = await this.getFromDB();
+    return !dbContract;
+  }
+
+  deployArgs () : TDeployArgs {
+    return [];
+  }
+
+  getArtifact () {
+    return this.campaign.deployer.getContractArtifact(this.contractName);
+  }
+
+  buildDbObject (hhContract : Contract, implAddress : string | null) : IContractDbObject {
```

Comment:
Consider making `type Maybe<T> = T | undefined` for uses like with implAddress

Follow up question, does this value have to be `null` for a specific reason?

Reply:
`null` is specifically here instead of `undefined`.
undefined type is too general and can be that because of ton of different things that can happen in code.
`null` is a more specific type here because it signifies that dbAdapter did not find anything.